drivers/cc1200/cc1200-rxtx.c

```diff
@@ -0,0 +1,430 @@
+          __func__, left, to_send, fifo, dev->pkt_buf.pos);
+
+    /* Flush TX  and calibrate Crystal */
+    cc1200_strobe(dev, CC1200_SFTX);
+    cc1200_strobe(dev, CC1200_SFSTXON);
+
+    /* Write packet into TX FIFO */
+    cc1200_writeburst_reg(dev, CC1200_TXFIFO, ((char *)pkt) + dev->pkt_buf.pos, to_send);
+    dev->pkt_buf.pos += to_send;
+
+    if (left == size) {
+        /* Switch to TX mode */
+        cc1200_strobe(dev, CC1200_STX);
+        //xtimer_usleep(200);
+    }
+
+    if (to_send < left) {
+        /* set GDO2 to 0x2 -> will deassert at TX FIFO below threshold */
+        cc1200_write_reg(dev, CC1200_IOCFG2, CC1200_CFG_TXFIFO_THR);
+        gpio_irq_enable(dev->params.int_pin);
+    }
+    else {
+        /* set GDO2 to 0x6 -> pin 2 will deassert at end of TX packet */
+        cc1200_write_reg(dev, CC1200_IOCFG2, CC1200_CFG_PKT_SYNC_RXTX);
+        gpio_irq_enable(dev->params.int_pin);
+    }
+}
+
+void cc1200_isr_handler(cc1200_t *dev, void (*callback)(void *), void *arg)
+{
+    uint8_t rxbytes;
+
+    switch (dev->radio_state) {
+        case RADIO_RX:
+            rxbytes = cc1200_read_reg(dev, CC1200_NUM_RXBYTES);
+            //DEBUG("radio rx\n");
+            if (gpio_read(dev->params.int_pin) | (rxbytes > 0)) {
+                //DEBUG("cc1200_isr_handler((): starting RX\n");
+                dev->isr_cb = callback;
+                dev->isr_cb_arg = arg;
+                _rx_start(dev);
+
+                /*
+                 * Don't know, why this waiting is necessary, but the driver
+                 * blocks without this. Alternatively, the delay can also be
+                 * located in  _netdev_cc1200_isr() or _isr(), but this here
+                 * is more specific. Also, the delay can be placed before or
+                 * after, but not in _rx_start()...
+                 */
+                xtimer_spin(xtimer_ticks_from_usec(800));
+            }
+            else {
+                DEBUG("cc1200_isr_handler((): isr handled too slow or falling edge trigger\n");
+                _rx_abort(dev);
+            }
+            break;
+        case RADIO_RX_BUSY:
+            //DEBUG("radio rx busy\n");
+            _rx_continue(dev, callback, arg);
+            break;
+        case RADIO_TX_BUSY:
+            //DEBUG("radio tx busy\n");
+            if (!gpio_read(dev->params.int_pin)) {
+                _tx_continue(dev);
+            }
+            else {
+                //DEBUG("cc1200_isr_handler() RADIO_TX_BUSY + GDO2\n");
+            }
+            break;
+        default:
+            DEBUG("%s:%s:%u: unhandled mode\n", RIOT_FILE_RELATIVE,
+                  __func__, __LINE__);
+            break;
+    }
+}
+
+int cc1200_send(cc1200_t *dev, const struct iovec *vector, unsigned count)
+{
+    DEBUG("%s:%u\n", __func__, __LINE__);
+
+    switch (dev->radio_state) {
+        case RADIO_RX_BUSY:
+            DEBUG("%s: rx busy...\n", __func__);
+            return -EAGAIN;
+        case RADIO_TX_BUSY:
```

Comment:
The implementation of this case looks a bit  unsteady to me. I have the impression the the state machine is not handeled correctly, or it is just my lack of knowledge about this device. Did you test subsequent transmissions in a row? E.g., sending IPv6 packets with 6LoWPAN fragmentation?

Reply:
I have written a stress test, which would send a packet without fragmentation every 0.5 seconds. In case a transmission received no `ACK` it would wait for 1 second, increasing with every package not acknowledged until a maximum wait time of 1 minute or until 200 packages have not been correctly transmitted. The program was running for 3 days until the host-machine unfortunately restarted, killing the power to my Firefly. Fragmented packages were also transmitted without a problem. Other than that, I have not done any testing, but would be nice if anyone wants to give it a go.

Reply:
I would do testing once we got the code in shape and you give me a "go"